File: src/llm/usage.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { appendFileSync, mkdirSync, readFileSync } from "node:fs";
import { resolve } from "node:path";

// One canonical wide event per run lands here as JSONL — `nr usage` sums it.
const CACHE_DIR = resolve(process.cwd(), ".cache");
const USAGE_LOG = resolve(CACHE_DIR, "usage.jsonl");

export type CallSite =
  | "research"
  | "critique"
  | "revise"
  | "re-critique"
  | "discover-plan"
  | "discover-rationale";

export interface CallUsage {
  callSite: CallSite;
  model: string;
  inputTokens: number;
  outputTokens: number;
  cachedInputTokens: number; // input tokens served from cache (W2)
  costUsd: number; // real $ from OpenRouter usage accounting
  durationMs: number;
}

export interface RunEvent {
  ts: string;
  kind: "research" | "discover";
  subject: string; // symbol or goal
  mode: string; // thorough | cheap | …
  cached: boolean; // whole run served from result cache (W3) → zero LLM calls
  revised?: boolean;
  calls: CallUsage[];
  totals: {
    calls: number;
    inputTokens: number;
    outputTokens: number;
    cachedInputTokens: number;
    costUsd: number;
    durationMs: number;
  };
  node: string;
}

interface RunContext {
  kind: RunEvent["kind"];
  subject: string;
  mode: string;
  calls: CallUsage[];
  extra: { cached?: boolean; revised?: boolean };
}

const store = new AsyncLocalStorage<RunContext>();

// Minimal shape we read off a generateObject/generateText result.
interface TrackableResult {
  usage?: {
    inputTokens?: number;
    outputTokens?: number;
    cachedInputTokens?: number;
  };
  providerMetadata?: Record<string, unknown>;
  response?: { modelId?: string };
}

function num(v: unknown): number {
  return typeof v === "number" && Number.isFinite(v) ? v : 0;
}

function extract(callSite: CallSite, res: TrackableResult, durationMs: number): CallUsage {
  // OpenRouter usage accounting (authoritative for $ and cache hits).
  const or = (res.providerMetadata?.openrouter as { usage?: Record<string, unknown> } | undefined)
    ?.usage;
  const cachedFromOr = num(
    (or?.promptTokensDetails as { cachedTokens?: number } | undefined)?.cachedTokens,
  );

  return {
    callSite,
    model: res.response?.modelId ?? "unknown",
    inputTokens: num(res.usage?.inputTokens) || num(or?.promptTokens),
    outputTokens: num(res.usage?.outputTokens) || num(or?.completionTokens),
    cachedInputTokens: num(res.usage?.cachedInputTokens) || cachedFromOr,
    costUsd: num(or?.cost),
    durationMs,
  };
}

// Time an LLM call, record its usage into the active run (if any), return the
// result unchanged. Call sites stay one line: `await tracked("research", () => generateObject(...))`.
export async function tracked<R extends TrackableResult>(
  callSite: CallSite,
  run: () => Promise<R>,
): Promise<R> {
  const start = Date.now();
  const res = await run();
  const ctx = store.getStore();
  if (ctx) ctx.calls.push(extract(callSite, res, Date.now() - start));
  return res;
}

// Add run-level facts that aren't tied to one call (revised?, cache hit?).
export function annotateRun(fields: { cached?: boolean; revised?: boolean }): void {
  const ctx = store.getStore();
  if (ctx) Object.assign(ctx.extra, fields);
}

function emit(ctx: RunContext): RunEvent {
  const totals = ctx.calls.reduce(
    (t, c) => ({
      calls: t.calls + 1,
      inputTokens: t.inputTokens + c.inputTokens,
      outputTokens: t.outputTokens + c.outputTokens,
      cachedInputTokens: t.cachedInputTokens + c.cachedInputTokens,
      costUsd: t.costUsd + c.costUsd,
      durationMs: t.durationMs + c.durationMs,
    }),
    { calls: 0, inputTokens: 0, outputTokens: 0, cachedInputTokens: 0, costUsd: 0, durationMs: 0 },
  );

  const event: RunEvent = {
    ts: new Date().toISOString(),
    kind: ctx.kind,
    subject: ctx.subject,
    mode: ctx.mode,
    cached: ctx.extra.cached ?? false,
    calls: ctx.calls,
    totals,
    node: process.version,
  };
  if (ctx.extra.revised !== undefined) event.revised = ctx.extra.revised;

  try {
    mkdirSync(CACHE_DIR, { recursive: true });
    appendFileSync(USAGE_LOG, JSON.stringify(event) + "\n");
  } catch {
    // Never let logging break a run.
  }
  return event;
}

// Run `fn` inside a usage-tracking context; emit one wide event when it settles.
// Returns fn's result; the emitted event is available via the optional onEvent hook.
export async function withRun<T>(
  meta: { kind: RunEvent["kind"]; subject: string; mode?: string },
  fn: () => Promise<T>,
  onEvent?: (event: RunEvent) => void,
): Promise<T> {
  const ctx: RunContext = {
    kind: meta.kind,
    subject: meta.subject,
    mode: meta.mode ?? "thorough",
    calls: [],
    extra: {},
  };
  try {
    return await store.run(ctx, fn);
  } finally {
    onEvent?.(emit(ctx));
  }
}

export function formatEvent(e: RunEvent): string {
  const t = e.totals;
  const head = `[usage] ${e.kind} ${e.subject} · mode=${e.mode}${e.cached ? " · CACHE HIT (0 calls)" : ""}${
    e.revised ? " · revised" : ""
  }`;
  const lines = [head];
  for (const c of e.calls) {
    lines.push(
      `  ${c.callSite.padEnd(16)} ${c.model.padEnd(28)} in=${c.inputTokens} (cached ${c.cachedInputTokens}) out=${c.outputTokens} $${c.costUsd.toFixed(5)} ${c.durationMs}ms`,
    );
  }
  lines.push(
    `  TOTAL  calls=${t.calls} in=${t.inputTokens} (cached ${t.cachedInputTokens}) out=${t.outputTokens} $${t.costUsd.toFixed(5)} ${t.durationMs}ms`,
  );
  return lines.join("\n");
}

// --- `nr usage` rollup -------------------------------------------------------

export function readEvents(): RunEvent[] {
  let raw: string;
  try {
    raw = readFileSync(USAGE_LOG, "utf8");
  } catch {
    return [];
  }
  const events: RunEvent[] = [];
  for (const line of raw.split("\n")) {
    if (!line.trim()) continue;
    try {
      events.push(JSON.parse(line) as RunEvent);
    } catch {
      // skip malformed line
    }
  }
  return events;
}

export function formatRollup(limit = 20): string {
  const events = readEvents();
  if (events.length === 0) return "No usage recorded yet. Run `nr research <SYMBOL>` first.";

  const recent = events.slice(-limit);
  const lines: string[] = [];
  lines.push(`USAGE — last ${recent.length} of ${events.length} runs`);
  lines.push("=".repeat(72));
  for (const e of recent) {
    const t = e.totals;
    lines.push(
      `${e.ts}  ${e.kind.padEnd(8)} ${e.subject.slice(0, 22).padEnd(22)} ${e.mode.padEnd(8)} ` +
        `calls=${t.calls} in=${t.inputTokens}(c${t.cachedInputTokens}) out=${t.outputTokens} $${t.costUsd.toFixed(5)}${e.cached ? " [cache]" : ""}`,
    );
  }
  lines.push("-".repeat(72));
  const sum = events.reduce(
    (s, e) => ({
      cost: s.cost + e.totals.costUsd,
      input: s.input + e.totals.inputTokens,
      cached: s.cached + e.totals.cachedInputTokens,
      output: s.output + e.totals.outputTokens,
    }),
    { cost: 0, input: 0, cached: 0, output: 0 },
  );
  lines.push(
    `ALL ${events.length} runs:  $${sum.cost.toFixed(4)}  in=${sum.input} (cached ${sum.cached}, ${
      sum.input ? ((sum.cached / sum.input) * 100).toFixed(1) : "0"
    }%)  out=${sum.output}`,
  );
  return lines.join("\n");
}

File: app/llm/usage.py
"""Wide-event usage tracking. One canonical JSONL event per run; `stocks usage`
sums it. Port of the original llm/usage.ts (AsyncLocalStorage -> contextvars).
"""

from __future__ import annotations

import asyncio
import json
import logging
import random
import sys
import time
from contextlib import asynccontextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any, AsyncIterator, Literal

from pydantic_ai.exceptions import ModelHTTPError

from ..config import CACHE_DIR

USAGE_LOG = CACHE_DIR / "usage.jsonl"
logger = logging.getLogger(__name__)
_RETRYABLE_STATUS_CODES = {429, 502, 503, 529}

CallSite = Literal[
    "research", "critique", "revise", "re-critique", "discover-plan", "discover-rationale"
]


@dataclass
class CallUsage:
    call_site: str
    model: str
    input_tokens: int
    output_tokens: int
    cached_input_tokens: int  # input tokens served from cache
    cost_usd: float  # real $ from OpenRouter usage accounting (best-effort)
    duration_ms: int


@dataclass
class RunContext:
    kind: str  # research | discover
    subject: str  # symbol or goal
    mode: str  # thorough | cheap | …
    calls: list[CallUsage] = field(default_factory=list)
    extra: dict[str, Any] = field(default_factory=dict)


_store: ContextVar[RunContext | None] = ContextVar("usage_run", default=None)


def _num(v: Any) -> float:
    return float(v) if isinstance(v, (int, float)) else 0.0


def _cost(result: Any) -> float:
    """OpenRouter's USD cost isn't a first-class field in pydantic-ai's RunUsage,
    so dig it out of the usage details / provider_details where it lands."""
    details = getattr(result.usage, "details", None) or {}
    for k in ("cost", "total_cost", "cost_usd"):
        if isinstance(details, dict) and k in details:
            return _num(details[k])
    pd = getattr(result.response, "provider_details", None)
    if isinstance(pd, dict):
        if "cost" in pd:
            return _num(pd["cost"])
        usage = pd.get("usage")
        if isinstance(usage, dict) and "cost" in usage:
            return _num(usage["cost"])
    return 0.0


def record(call_site: str, result: Any, duration_ms: int) -> None:
    """Record one LLM call's usage into the active run (if any)."""
    ctx = _store.get()
    if ctx is None:
        return
    u = result.usage
    ctx.calls.append(
        CallUsage(
            call_site=call_site,
            model=getattr(result.response, "model_name", None) or "unknown",
            input_tokens=int(_num(u.input_tokens)),
            output_tokens=int(_num(u.output_tokens)),
            cached_input_tokens=int(_num(u.cache_read_tokens)),
            cost_usd=_cost(result),
            duration_ms=duration_ms,
        )
    )


def _settings_for_attempt(model_settings: Any, attempt: int, max_attempts: int) -> Any:
    if attempt != max_attempts or not isinstance(model_settings, dict):
        return model_settings

    provider = model_settings.get("openrouter_provider")
    if not isinstance(provider, dict) or provider.get("allow_fallbacks") is not False:
        return model_settings

    copied = dict(model_settings)
    copied_provider = dict(provider)
    copied_provider.pop("allow_fallbacks", None)
    copied["openrouter_provider"] = copied_provider
    return copied


async def run_tracked(
    call_site: str, agent: Any, user_prompt: Any, model_settings: Any, model: Any = None
) -> Any:
    """Time an agent run, record its usage, return the result unchanged.

    ``model`` optionally overrides the agent's bound model (the critic chain
    swaps premium/workhorse per call)."""
    max_attempts = 3
    for attempt in range(1, max_attempts + 1):
        kwargs: dict[str, Any] = {
            "model_settings": _settings_for_attempt(model_settings, attempt, max_attempts)
        }
        if model is not None:
            kwargs["model"] = model
        start = time.monotonic()
        try:
            result = await agent.run(user_prompt, **kwargs)
        except ModelHTTPError as exc:
            if exc.status_code not in _RETRYABLE_STATUS_CODES or attempt == max_attempts:
                raise
            logger.warning(
                "LLM call failed with retryable HTTP status",
                extra={"call_site": call_site, "status": exc.status_code, "attempt": attempt},
            )
            await asyncio.sleep(2 * attempt + random.uniform(0, 1))
            continue
        record(call_site, result, int((time.monotonic() - start) * 1000))
        return result

    raise RuntimeError("unreachable")


def annotate_run(**fields: Any) -> None:
    ctx = _store.get()
    if ctx is not None:
        ctx.extra.update(fields)


def _totals(calls: list[CallUsage]) -> dict[str, Any]:
    return {
        "calls": len(calls),
        "input_tokens": sum(c.input_tokens for c in calls),
        "output_tokens": sum(c.output_tokens for c in calls),
        "cached_input_tokens": sum(c.cached_input_tokens for c in calls),
        "cost_usd": sum(c.cost_usd for c in calls),
        "duration_ms": sum(c.duration_ms for c in calls),
    }


def _emit(ctx: RunContext) -> dict[str, Any]:
    event = {
        "ts": datetime.now(UTC).isoformat(),
        "kind": ctx.kind,
        "subject": ctx.subject,
        "mode": ctx.mode,
        "cached": ctx.extra.get("cached", False),
        "calls": [c.__dict__ for c in ctx.calls],
        "totals": _totals(ctx.calls),
        "python": sys.version.split()[0],
    }
    if "revised" in ctx.extra:
        event["revised"] = ctx.extra["revised"]
    try:
        CACHE_DIR.mkdir(parents=True, exist_ok=True)
        with USAGE_LOG.open("a") as f:
            f.write(json.dumps(event) + "\n")
    except OSError:
        pass  # never let logging break a run
    return event


@asynccontextmanager
async def with_run(kind: str, subject: str, mode: str = "thorough") -> AsyncIterator[RunContext]:
    """Run inside a usage-tracking context; emit one wide event when it settles.

    The emitted event is attached to ``ctx.extra['_event']`` for callers that
    want to print it.
    """
    ctx = RunContext(kind=kind, subject=subject, mode=mode)
    token = _store.set(ctx)
    try:
        yield ctx
    finally:
        ctx.extra["_event"] = _emit(ctx)
        _store.reset(token)


# --- Formatting -------------------------------------------------------------


def format_event(e: dict[str, Any]) -> str:
    t = e["totals"]
    head = (
        f"[usage] {e['kind']} {e['subject']} · mode={e['mode']}"
        f"{' · CACHE HIT (0 calls)' if e.get('cached') else ''}"
        f"{' · revised' if e.get('revised') else ''}"
    )
    lines = [head]
    for c in e["calls"]:
        lines.append(
            f"  {c['call_site']:<16} {c['model']:<28} "
            f"in={c['input_tokens']} (cached {c['cached_input_tokens']}) "
            f"out={c['output_tokens']} ${c['cost_usd']:.5f} {c['duration_ms']}ms"
        )
    lines.append(
        f"  TOTAL  calls={t['calls']} in={t['input_tokens']} "
        f"(cached {t['cached_input_tokens']}) out={t['output_tokens']} "
        f"${t['cost_usd']:.5f} {t['duration_ms']}ms"
    )
    return "\n".join(lines)


def read_events() -> list[dict[str, Any]]:
    try:
        raw = USAGE_LOG.read_text()
    except OSError:
        return []
    events = []
    for line in raw.splitlines():
        if not line.strip():
            continue
        try:
            events.append(json.loads(line))
        except ValueError:
            pass
    return events


def format_rollup(limit: int = 20) -> str:
    events = read_events()
    if not events:
        return "No usage recorded yet. Run `stocks research <SYMBOL>` first."

    recent = events[-limit:]
    lines = [f"USAGE — last {len(recent)} of {len(events)} runs", "=" * 72]
    for e in recent:
        t = e["totals"]
        lines.append(
            f"{e['ts']}  {e['kind']:<8} {e['subject'][:22]:<22} {e['mode']:<8} "
            f"calls={t['calls']} in={t['input_tokens']}(c{t['cached_input_tokens']}) "
            f"out={t['output_tokens']} ${t['cost_usd']:.5f}"
            f"{' [cache]' if e.get('cached') else ''}"
        )
    lines.append("-" * 72)
    cost = sum(e["totals"]["cost_usd"] for e in events)
    inp = sum(e["totals"]["input_tokens"] for e in events)
    cached = sum(e["totals"]["cached_input_tokens"] for e in events)
    out = sum(e["totals"]["output_tokens"] for e in events)
    pct = (cached / inp * 100) if inp else 0
    lines.append(
        f"ALL {len(events)} runs:  ${cost:.4f}  "
        f"in={inp} (cached {cached}, {pct:.1f}%)  out={out}"
    )
    return "\n".join(lines)
